java/vector/src/main/java/org/apache/arrow/vector/table/BaseTable.java

```diff
@@ -0,0 +1,387 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.vector.table;
+
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import org.apache.arrow.util.AutoCloseables;
+import org.apache.arrow.util.Preconditions;
+import org.apache.arrow.vector.FieldVector;
+import org.apache.arrow.vector.ValueVector;
+import org.apache.arrow.vector.VectorSchemaRoot;
+import org.apache.arrow.vector.complex.reader.FieldReader;
+import org.apache.arrow.vector.dictionary.Dictionary;
+import org.apache.arrow.vector.dictionary.DictionaryEncoder;
+import org.apache.arrow.vector.dictionary.DictionaryProvider;
+import org.apache.arrow.vector.types.pojo.Field;
+import org.apache.arrow.vector.types.pojo.Schema;
+import org.apache.arrow.vector.util.TransferPair;
+
+/**
+ * Abstract base class for Table.
+ * <p>
+ * This API is EXPERIMENTAL.
+ */
+public abstract class BaseTable implements AutoCloseable {
+
+  /** The field vectors holding the data in this table. */
+  protected final List<FieldVector> fieldVectors;
+
+  /**
+   * An optional DictionaryProvider. One must be present if any vector in the table is dictionary
+   * encoded.
+   */
+  protected DictionaryProvider dictionaryProvider;
+
+  /** A map of Fields to FieldVectors used to select Fields. */
+  protected final Map<Field, FieldVector> fieldVectorsMap = new LinkedHashMap<>();
+
+  /** The schema for the table. */
+  protected Schema schema;
+
+  /** The number of rows of data in the table; not necessarily the same as the table row capacity. */
+  protected int rowCount;
```

Comment:
Same here - I think it'd be best to use `long` everywhere for indexing from the get-go if we do anticipate ChunkedArray support (things currently use int because a single allocation can't hold more elements than that, but ChunkedArray would remove that restriction)

Reply:
Ok